File: frontend/src/api/dashboards.ts
import { apiClient } from './client';

export interface Widget {
  id: number;
  dashboard_id: number;
  widget_type: string;
  title: string;
  config: Record<string, unknown>;
  local_filters: Record<string, unknown> | null;
  use_global_filters: boolean;
  grid_x: number;
  grid_y: number;
  grid_width: number;
  grid_height: number;
}

export interface Dashboard {
  id: number;
  virtual_dataset_id: number;
  name: string;
  description: string | null;
  global_filters: Record<string, unknown>;
  applied_slice_id: number | null;
  layout: unknown[];
  created_by: number;
  created_at: string;
  widgets: Widget[];
}

export interface DashboardBrief {
  id: number;
  name: string;
  created_by: number;
  created_at: string;
}

export interface WidgetCreatePayload {
  widget_type: string;
  title: string;
  config?: Record<string, unknown>;
  grid_width?: number;
  grid_height?: number;
}

export async function listDashboards(
  projectId: number,
  vdId: number
): Promise<{ items: DashboardBrief[]; total: number }> {
  const { data } = await apiClient.get<{ items: DashboardBrief[]; total: number }>(
    `/projects/${projectId}/virtual-datasets/${vdId}/dashboards`
  );
  return data;
}

export async function createDashboard(
  projectId: number,
  vdId: number,
  payload: { name: string; description?: string | null }
): Promise<Dashboard> {
  const { data } = await apiClient.post<Dashboard>(
    `/projects/${projectId}/virtual-datasets/${vdId}/dashboards`,
    payload
  );
  return data;
}

export async function getDashboard(dashboardId: number): Promise<Dashboard> {
  const { data } = await apiClient.get<Dashboard>(`/dashboards/${dashboardId}`);
  return data;
}

export async function deleteDashboard(dashboardId: number): Promise<void> {
  await apiClient.delete(`/dashboards/${dashboardId}`);
}

export async function updateDashboard(
  dashboardId: number,
  payload: {
    name?: string;
    description?: string | null;
    global_filters?: Record<string, unknown>;
  }
): Promise<Dashboard> {
  const { data } = await apiClient.patch<Dashboard>(
    `/dashboards/${dashboardId}`,
    payload
  );
  return data;
}

export async function addWidget(
  dashboardId: number,
  payload: WidgetCreatePayload
): Promise<Widget> {
  const { data } = await apiClient.post<Widget>(
    `/dashboards/${dashboardId}/widgets`,
    payload
  );
  return data;
}

export async function deleteWidget(widgetId: number): Promise<void> {
  await apiClient.delete(`/widgets/${widgetId}`);
}

export interface WidgetLayoutItem {
  id: number;
  grid_x: number;
  grid_y: number;
  grid_width: number;
  grid_height: number;
}

export async function updateDashboardLayout(
  dashboardId: number,
  items: WidgetLayoutItem[]
): Promise<Dashboard> {
  const { data } = await apiClient.patch<Dashboard>(
    `/dashboards/${dashboardId}/widgets/layout`,
    { widgets: items }
  );
  return data;
}

export async function getWidgetData(
  widgetId: number
): Promise<Record<string, unknown>> {
  const { data } = await apiClient.get<Record<string, unknown>>(
    `/widgets/${widgetId}/data`
  );
  return data;
}

File: backend/app/api/v1/dashboards.py
from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from typing import Any

from app.core.exceptions import (
    BusinessRuleError,
    EntityNotFoundError,
    PermissionDeniedError,
)
from app.db.models.dashboards import Dashboard, DashboardWidget
from app.db.models.users import User
from app.db.session import get_db
from app.schemas.dashboards import (
    DashboardBrief,
    DashboardCreate,
    DashboardLayoutUpdate,
    DashboardList,
    DashboardResponse,
    DashboardUpdate,
    WidgetCreate,
    WidgetResponse,
    WidgetUpdate,
)
from app.services import dashboard_service, widget_data_service

router = APIRouter(tags=["Дашборды"])


def _to_response(
    dashboard: Dashboard, widgets: list[DashboardWidget]
) -> DashboardResponse:
    return DashboardResponse(
        id=dashboard.id,
        virtual_dataset_id=dashboard.virtual_dataset_id,
        name=dashboard.name,
        description=dashboard.description,
        global_filters=dashboard.global_filters,
        applied_slice_id=dashboard.applied_slice_id,
        layout=dashboard.layout,
        template_kind=dashboard.template_kind,
        created_by=dashboard.created_by,
        created_at=dashboard.created_at,
        widgets=[WidgetResponse.model_validate(w) for w in widgets],
    )


@router.post(
    "/projects/{project_id}/virtual-datasets/{vd_id}/dashboards",
    response_model=DashboardResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_dashboard(
    project_id: int,
    vd_id: int,
    payload: DashboardCreate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> DashboardResponse:
    try:
        dashboard = await dashboard_service.create_dashboard(
            db, vd_id, payload, user, request
        )
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    return _to_response(dashboard, [])


@router.get(
    "/projects/{project_id}/virtual-datasets/{vd_id}/dashboards",
    response_model=DashboardList,
)
async def list_dashboards(
    project_id: int,
    vd_id: int,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
) -> DashboardList:
    items, total = await dashboard_service.list_dashboards(db, vd_id)
    return DashboardList(
        items=[DashboardBrief.model_validate(d) for d in items], total=total
    )


@router.get("/dashboards/{dashboard_id}", response_model=DashboardResponse)
async def get_dashboard(
    dashboard_id: int,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
) -> DashboardResponse:
    try:
        dashboard, widgets = await dashboard_service.get_dashboard(db, dashboard_id)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    return _to_response(dashboard, widgets)


@router.patch("/dashboards/{dashboard_id}", response_model=DashboardResponse)
async def update_dashboard(
    dashboard_id: int,
    payload: DashboardUpdate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> DashboardResponse:
    try:
        await dashboard_service.update_dashboard(db, dashboard_id, payload, user, request)
        dashboard, widgets = await dashboard_service.get_dashboard(db, dashboard_id)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    return _to_response(dashboard, widgets)


@router.patch(
    "/dashboards/{dashboard_id}/widgets/layout",
    response_model=DashboardResponse,
)
async def update_widget_layouts(
    dashboard_id: int,
    payload: DashboardLayoutUpdate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> DashboardResponse:
    """Batch-обновление координат виджетов после drag/resize."""
    try:
        await dashboard_service.update_widget_layouts(
            db, dashboard_id, payload, user, request
        )
        dashboard, widgets = await dashboard_service.get_dashboard(db, dashboard_id)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    except BusinessRuleError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, str(exc)) from exc
    return _to_response(dashboard, widgets)


@router.delete("/dashboards/{dashboard_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_dashboard(
    dashboard_id: int,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> None:
    try:
        await dashboard_service.delete_dashboard(db, dashboard_id, user, request)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc


@router.post(
    "/dashboards/{dashboard_id}/duplicate",
    response_model=DashboardResponse,
    status_code=status.HTTP_201_CREATED,
)
async def duplicate_dashboard(
    dashboard_id: int,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> DashboardResponse:
    try:
        copy = await dashboard_service.duplicate_dashboard(db, dashboard_id, user, request)
        dashboard, widgets = await dashboard_service.get_dashboard(db, copy.id)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    return _to_response(dashboard, widgets)


@router.post(
    "/dashboards/{dashboard_id}/widgets",
    response_model=WidgetResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_widget(
    dashboard_id: int,
    payload: WidgetCreate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> WidgetResponse:
    try:
        widget = await dashboard_service.add_widget(
            db, dashboard_id, payload, user, request
        )
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    return WidgetResponse.model_validate(widget)


@router.get("/widgets/{widget_id}/data")
async def get_widget_data(
    widget_id: int,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """Рассчитанные данные для отрисовки виджета (форма зависит от типа)."""
    try:
        widget = await dashboard_service.get_widget(db, widget_id)
        return await widget_data_service.compute_widget_data(db, widget)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except BusinessRuleError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, str(exc)) from exc


@router.patch("/widgets/{widget_id}", response_model=WidgetResponse)
async def update_widget(
    widget_id: int,
    payload: WidgetUpdate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> WidgetResponse:
    try:
        widget = await dashboard_service.update_widget(
            db, widget_id, payload, user, request
        )
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
    return WidgetResponse.model_validate(widget)


@router.delete("/widgets/{widget_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_widget(
    widget_id: int,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> None:
    try:
        await dashboard_service.delete_widget(db, widget_id, user, request)
    except EntityNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except PermissionDeniedError as exc:
        raise HTTPException(status.HTTP_403_FORBIDDEN, str(exc)) from exc
